integrations/datadog/client.py

```diff
@@ -233,6 +296,117 @@ async def get_slo_history(
             )
             return {}
 
+    def _create_query_with_values(
+        self, metric_query: str, variable_values: dict[str, str]
+    ) -> str:
+        """
+        Creates a Datadog query by substituting template variables with their values.
+
+        Args:
+            metric_query (str): The original query string containing template variables (e.g., "avg:container.cpu.usage{$service}").
+            variable_values (dict): A dictionary mapping variable names to their values (e.g., {"service": "svc_name", "env": "prod"}).
+
+        Returns:
+            str: The modified query with variables replaced by their values.
+        """
+        logger.debug(f"Creating query with values: {metric_query} | {variable_values}")
+        for variable_name, value in variable_values.items():
+            metric_query = metric_query.replace(
+                f"${variable_name}", f"{variable_name}:{value}"
+            )
+        return metric_query
+
+    def extract_metric_name(self, query_string):
+        """Extracts the Datadog metric name from a query string.
+
+        Args:
+            query_string: The Datadog query string.
+
+        Returns:
+            The metric name, or None if no metric name is found.
+
+        Examples:
+            - "sum:container.memory.usage" -> "container.memory.usage"
+            - "avg:system.cpu.used{service:my-service}" -> "system.cpu.used"
+        """
+        match = re.search(r"(?:\w+:)?(\w+\.\w+\.\w+)", query_string)
+        if match:
+            return match.group(1)
+        else:
+            return None
+
+    async def get_single_service(self, service_id: str) -> dict[str, Any] | None:
+        if not service_id:
+            return None
+        url = f"{self.api_url}/api/v2/services/definitions/{service_id}"
+        return await self._send_api_request(url)
+
+    async def get_metric_metadata(self, metric: str) -> dict[str, Any] | None:
+        url = f"{self.api_url}/api/v1/metrics/{metric}"
+        return await self._send_api_request(url)
+
+    async def get_metrics(
+        self, query: str, env: str = "prod", service: str = "*"
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        """
+        Fetches metrics for specified services and environment.
+
+        Args:
+            query (str): The Datadog query string to execute.
+            env (str): The environment to filter by (default: "prod").
+            service (str): The service ID to filter by, or "*" to fetch metrics for all services.
+
+        Yields:
+            AsyncGenerator[list[dict[str, Any]], None]: A list of metrics for each service.
+        """
+        url = f"{self.api_url}/api/v1/metrics"
+
+        metrics = []
+
+        services = []
+        if service == "*":
+            # get all services
+            async for service_list in self.get_services():
+                services.extend(service_list)
```

Comment:
should we actually wait for all the services to start fetching service metrics? why not after receiving a batch starting querying its metrics?